src/lib.rs

```diff
@@ -1467,3 +1433,96 @@ impl Disposition {
         }
     }
 }
+
+// Request helper type for NATS request
+struct Request {
+    reply_to: Subject,
+    wrapped_client: Arc<Mutex<SyncClient>>,
+}
+
+impl Request {
+    async fn new(wrapped_client: Arc<Mutex<SyncClient>>) -> Result<Self> {
+        let inbox_uuid = Uuid::new_v4();
+
+        let client = wrapped_client.lock().await;
+        let request_inbox = inbox_uuid.to_simple();
+        let reply_to: Subject = format!(
+            "{}.{}.{}",
+            util::INBOX_PREFIX,
+            client.request_base_inbox,
+            request_inbox
+        )
+        .parse()?;
+
+        Ok(Self {
+            reply_to,
+            wrapped_client: wrapped_client.clone(),
+        })
+    }
+
+    async fn call(
+        &mut self,
+        subject: &Subject,
+        payload: &[u8],
+        duration: Option<Duration>,
+    ) -> Result<Msg> {
+        let mut rx = {
+            let mut client = self.wrapped_client.lock().await;
+
+            // Only subscribe to the wildcard subscription when requested once!
+            if client.request_wildcard_subscription.is_none() {
+                let global_reply_to =
+                    format!("{}.{}.*", util::INBOX_PREFIX, client.request_base_inbox).parse()?;
+                let (sid, rx) = client.subscribe(&global_reply_to, 1024).await?;
+                client.request_wildcard_subscription = Some(sid);
+
+                // Spawn the task that watches the request wildcard receiver.
+                tokio::spawn(SyncClient::request_wildcard_handler(
+                    self.wrapped_client.clone(),
+                    rx,
+                ));
+            }
+
+            let (tx, rx) = mpsc::channel(1);
+            client
+                .request_inbox_mapping
+                .insert(self.reply_to.clone(), tx);
+            client
+                .publish_with_reply(subject, &self.reply_to, payload)
+                .await?;
+
+            rx
+        };
+
+        // Use a timeout duration if it was provided by the caller.
+        let next_message = match duration {
+            Some(duration) => tokio::time::timeout(duration, rx.next()).await?,
+            None => rx.next().await,
+        };
+
+        // Make sure we clean up on error (don't leave a dangling request
+        // inbox mapping reference. Adding an extra mutex here seems fine
+        // since this is the error path.
+        match next_message {
+            Some(response) => Ok(response),
+            None => Err(Error::NoResponse),
+        }
+    }
+}
+
+impl Drop for Request {
+    // When the request is dropped, ensure the reply_to subject is removed
+    // from the client request inbox mapping.
+    // NOTE: This is a blocking async block, but the only thing it blocks on
+    // is the client's mutex, which should be fine.
```

Comment:
Would you mind also adding a note that when/if [async drop](https://boats.gitlab.io/blog/post/poll-drop/) becomes a thing this will no longer need to be blocking.